pinot-core/src/main/java/com/linkedin/pinot/core/realtime/segment/ConsumingSegmentAssignmentStrategy.java

```diff
@@ -0,0 +1,55 @@
+/**
+ * Copyright (C) 2014-2016 LinkedIn Corp. (pinot-core@linkedin.com)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.linkedin.pinot.core.realtime.segment;
+
+import com.linkedin.pinot.common.partition.PartitionAssignment;
+import com.linkedin.pinot.common.utils.LLCSegmentName;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+
+/**
+ * Assigns the given list of consuming segments onto instances based on given partition assignment
+ */
+public class ConsumingSegmentAssignmentStrategy implements RealtimeSegmentAssignmentStrategy {
+
+  /**
+   * Assigns new segments to instances by referring to the partition assignment
+   * @param newSegments segments to assign
+   * @param partitionAssignment partition assignment for the table to which the segments belong
+   * @return map of segment name to instances list
+   */
+  public Map<String, List<String>> assign(List<String> newSegments, PartitionAssignment partitionAssignment) {
+
+    Map<String, List<String>> segmentAssignment = new HashMap<>(newSegments.size());
+
+    for (String segmentName : newSegments) {
+      if (LLCSegmentName.isLowLevelConsumerSegmentName(segmentName)) {
+        LLCSegmentName llcSegmentName = new LLCSegmentName(segmentName);
+        int partitionId = llcSegmentName.getPartitionId();
+        List<String> instancesListForPartition = partitionAssignment.getInstancesListForPartition(String.valueOf(partitionId));
+        if (instancesListForPartition == null) {
+          throw new IllegalStateException("No partition assignment " + partitionId + " found for segment " + segmentName);
```

Comment:
better to throw a compile exception so the caller is forced to check. If it is new table access or rebalance, we can throw back the exception, but if it is on a dynamic change, we need to update a metric.